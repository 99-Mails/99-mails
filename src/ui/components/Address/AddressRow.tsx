import type { AddressID } from "@/domains/Address";
import {
  Tag,
  Flex,
  Text,
  IconButton,
  useClipboard,
  forwardRef,
  IconButtonProps,
  TextProps,
  TagProps,
  HStack,
  Spinner,
} from "@chakra-ui/react";
import { Suspense, useEffect, useMemo, useState } from "react";
import useDeleteAddressFromSession from "@/application/deleteAddressFromSession";
import { CopyIcon, DeleteIcon } from "@chakra-ui/icons";
import { Timer } from "../Timer";
import { FancyDate } from "@/lib/date";

const AddressRowButton = forwardRef<IconButtonProps, "button">((props, ref) => {
  return <IconButton ref={ref} {...props} />;
});

type AddressRowNameProps = TextProps & {
  name: string;
};

const AddressRowName = forwardRef<AddressRowNameProps, "div">((props, ref) => {
  const { name } = props;
  return (
    <Text ref={ref} fontSize="3xl" {...props}>
      {name}
    </Text>
  );
});

type AddressRowCountProps = TagProps & {
  count: number;
};

const AddressRowCount = forwardRef<AddressRowCountProps, "span">(
  (props, ref) => {
    const { count } = props;
    return (
      <Tag ref={ref} bg="gray.500" color="white" {...props}>
        {count}
      </Tag>
    );
  }
);

const AddressRow = ({
  address,
  count,
}: {
  address: AddressID;
  count: number;
}) => {
  const [disabled, disableAddress] = useState(false);
  const [expiresAt, setExpiresAt] = useState("");
  const { doDeleteAddress } = useDeleteAddressFromSession();
  const {
    onCopy: onClipboard,
    setValue,
    hasCopied,
  } = useClipboard("", { timeout: 1000 });

  useEffect(() => {
    function init() {
      setValue(address.address);
    }
    init();
    return () => setValue("");
  }, []);

  useEffect(() => {
    const timer = setTimeout(() => disableAddress(true), 10 * 60000);
    return () => clearTimeout(timer);
  }, []);

  useMemo(() => {
    setExpiresAt(new FancyDate().getDateWithMinutesFromNow(10));
  }, []);

  return (
    <Flex gap={2} alignItems={"center"} justifyContent="space-between">
      <AddressRowName
        as={disabled ? "del" : "p"}
        data-testid="email-address"
        name={address.address}
      />
      <HStack>
        <AddressRowButton
          data-testid="copy-address"
          isDisabled={disabled}
          onClick={onClipboard}
          size="sm"
          icon={<CopyIcon />}
          aria-label={""}
          colorScheme={hasCopied ? "green" : "gray"}
        />
        <AddressRowButton
          data-testid="delete-address"
          isDisabled={disabled}
          onClick={() => doDeleteAddress(address)}
          size="sm"
          icon={<DeleteIcon />}
          aria-label={""}
        />
        <AddressRowCount data-testid="email-count" count={count} />
        <Tag minW="60px">
          <Suspense fallback={<Spinner />}>
            <Timer expiresAt={expiresAt} isDisabled={disabled} />
          </Suspense>
        </Tag>
      </HStack>
    </Flex>
  );
};

export default AddressRow;
