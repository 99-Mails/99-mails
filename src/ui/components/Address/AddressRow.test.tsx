import { vi, it, expect } from "vitest";
import { renderWithContext, screen, userEvent } from "@/tests/test-utils";
import type { IStoreContext } from "@/services/store/store";
import { AddressRow } from "./AddressRow";
import type { AddressID } from "@/domains/Address";

let storeProps: IStoreContext;

let address: AddressID = {
  address: "asd@fs.te",
  id: "asdfasdfasdfsdfasdf",
};

beforeEach(() => {
  storeProps = {
    dispatch: vi.fn,
    state: {
      tempEmails: {
        sessionId: "123123",
        expiresAt: 0,
      },
    },
  };
});

it("should render correctly", () => {
  const { asFragment } = renderWithContext(
    <AddressRow address={address.address} id={address.id} />,
    { storeProps }
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should display the email address and it's count", () => {
  renderWithContext(<AddressRow address={address.address} id={address.id} />, {
    storeProps,
  });
  expect(screen.getByTestId("email-address")).toHaveTextContent("asd@fs.te");
});

it.skip("should display the email count", () => {
  renderWithContext(<AddressRow address={address.address} id={address.id} />, {
    storeProps,
  });
  expect(screen.getByTestId("email-count")).toHaveTextContent("2");
});

it("should be able to copy the email address to the clipboard", async () => {
  window.prompt = vi.fn();
  document.execCommand = vi.fn();

  const user = userEvent.setup();

  renderWithContext(<AddressRow address={address.address} id={address.id} />, {
    storeProps,
  });

  const copyButton = screen.getByTestId("copy-address");
  await user.click(copyButton);

  expect(document.execCommand).toHaveBeenCalledOnce();
  expect(document.execCommand).toHaveBeenCalledWith("copy");
  expect(window.prompt).toHaveBeenCalledWith(
    "Copy to clipboard: Ctrl+C, Enter",
    "asd@fs.te"
  );
});

// TODO: find a way to test it,
//       maybe check if it's not availabe anymore in the dom
// it("should be able to remove the email address from the session", async () => {
//   const user = userEvent.setup();

//   renderWithContext(<AddressRow address={address} count={2} />, { storeProps });

//   const deleteButton = screen.getByTestId("delete-address");
//   await user.click(deleteButton);

//   expect(true).toBe(true);
// });
