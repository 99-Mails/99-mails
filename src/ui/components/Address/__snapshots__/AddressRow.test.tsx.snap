// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render correctly 1`] = `
<DocumentFragment>
  <div
    class="css-1blblq8"
  >
    <p
      class="chakra-text css-1tapsqs"
      data-testid="email-address"
      name="asd@fs.te"
    >
      asd@fs.te
    </p>
    <div
      class="chakra-stack css-84zodg"
    >
      <button
        aria-label=""
        class="chakra-button css-p7mkev"
        data-testid="copy-address"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm3 4H8c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h11c1.1 0 2-.9 2-2V7c0-1.1-.9-2-2-2zm0 16H8V7h11v14z"
            fill="currentColor"
          />
        </svg>
      </button>
      <button
        aria-label=""
        class="chakra-button css-p7mkev"
        data-testid="delete-address"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="chakra-icon css-onkibi"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <g
            fill="currentColor"
          >
            <path
              d="M19.452 7.5H4.547a.5.5 0 00-.5.545l1.287 14.136A2 2 0 007.326 24h9.347a2 2 0 001.992-1.819L19.95 8.045a.5.5 0 00-.129-.382.5.5 0 00-.369-.163zm-9.2 13a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zm5 0a.75.75 0 01-1.5 0v-9a.75.75 0 011.5 0zM22 4h-4.75a.25.25 0 01-.25-.25V2.5A2.5 2.5 0 0014.5 0h-5A2.5 2.5 0 007 2.5v1.25a.25.25 0 01-.25.25H2a1 1 0 000 2h20a1 1 0 000-2zM9 3.75V2.5a.5.5 0 01.5-.5h5a.5.5 0 01.5.5v1.25a.25.25 0 01-.25.25h-5.5A.25.25 0 019 3.75z"
            />
          </g>
        </svg>
      </button>
      <span
        class="css-x780zy"
        count="2"
        data-testid="email-count"
      >
        2
      </span>
      <span
        class="css-isuppp"
      >
        <b
          class="chakra-text css-0"
          data-testid="timer"
        />
      </span>
    </div>
  </div>
</DocumentFragment>
`;
